File: EEntityCore.DB/EEntityCore.DB/Abstracts/SimpleTablePlugIn.vb
Imports System.Data
Imports ELibrary.Standard.VB.Objects
Imports EEntityCore.DB.Abstracts.All__DBs
Imports ELibrary.Standard.VB

Namespace Abstracts


    ''' <summary>
    ''' Simple Table with target row capability
    ''' </summary>
    ''' <remarks></remarks>
    Public MustInherit Class SimpleTablePlugIn

#Region "Enums & Consts"

        Public Enum RecordAccessibility
            ''' <summary>
            ''' Access table with SQL QUERIES and target record. reloadable with DBConn Valid
            ''' </summary>
            ''' <remarks></remarks>
            FULL_ACCESS
            ''' <summary>
            ''' Access table and target record via direct record (DataRow(s)). reloadable if dbConn is valid
            ''' </summary>
            ''' <remarks></remarks>
            PARTIAL_ACCESS
            ' '' ''' <summary>
            ' '' ''' Access table and target record via direcet record. reloadable
            ' '' ''' </summary>
            ' '' ''' <remarks></remarks>
            ' ''PARTIAL_RELOADABLE_ACCESS
            ' '' ''' <summary>
            ' '' ''' Access table and target record. NOT reloadable [Cached]
            ' '' ''' </summary>
            ' '' ''' <remarks></remarks>
            ' ''PARTIAL_FIXED_ACCESS
            ''' <summary>
            ''' Access target record only. NOT reloadable. No DBConn
            ''' </summary>
            ''' <remarks></remarks>
            SHALLOW_ACCESS

            ''' <summary>
            ''' Signifies EMPTY UNDECIDED CLASS
            ''' </summary>
            ''' <remarks></remarks>
            UNKNOWN

        End Enum


        Public Const DO__NOT____TARGET__ANY_ROWID As Int32 = 0

#End Region

#Region "Contructors"
        ''' <summary>
        ''' UNKNOWN ACCESS. Just sets table name only
        ''' </summary>
        ''' <param name="TableName"></param>
        ''' <remarks></remarks>
        Public Sub New(
                    ByVal TableName As String
                    )
            Me.____TableName = TableName
            Me._OpenAccessFor = RecordAccessibility.UNKNOWN
        End Sub


#Region "Full Access"

        '   Full Access means initial data is loaded directly from database, so DBConn MUST be provided

        ''' <summary>
        ''' FULL ACCESS. This will load data with dbconn
        ''' </summary>
        ''' <param name="DBConn"></param>
        ''' <param name="TableName"></param>
        ''' <remarks></remarks>
        Public Sub New(ByVal DBConn As All__DBs,
                      ByVal TableName As String
                     )
            Me.New(DBConn, TableName, DO__NOT____TARGET__ANY_ROWID)

        End Sub

        ''' <summary>
        ''' Full Access. This will load data with dbconn
        ''' </summary>
        ''' <param name="DBConn"></param>
        ''' <param name="pTableName"></param>
        ''' <param name="pTargettedRowID">Only works if the table contains a column named ID</param>
        ''' <remarks></remarks>
        Sub New(ByVal DBConn As All__DBs,
                ByVal pTableName As String,
                      ByVal pTargettedRowID As Int64)

            Me.New(DBConn, pTableName, pTargettedRowID:=pTargettedRowID, pSQL:="SELECT * FROM " & pTableName)

        End Sub

        ''' <summary>
        ''' Full Access. This will load data with dbconn
        ''' </summary>
        ''' <param name="pDBConn"></param>
        ''' <param name="pTableName"></param>
        ''' <param name="pSQL">Load table with your own SQL</param>
        ''' <remarks></remarks>
        Public Sub New(ByVal pDBConn As All__DBs,
                ByVal pTableName As String,
                ByVal pSQL As String)
            Me.New(pDBConn, pTableName, pSQL, DO__NOT____TARGET__ANY_ROWID)
        End Sub

        ''' <summary>
        ''' Full Access. This will load data with dbconn
        ''' </summary>
        ''' <param name="pDBConn"></param>
        ''' <param name="pTableName"></param>
        ''' <param name="pTargettedRowID"></param>
        ''' <param name="pSQL">Load table with your own SQL</param>
        ''' <remarks></remarks>
        Sub New(ByVal pDBConn As All__DBs,
                ByVal pTableName As String,
                ByVal pSQL As String,
                     ByVal pTargettedRowID As Int64)

            Me.New(pTableName)
            Me._OpenAccessFor = RecordAccessibility.FULL_ACCESS
            Me._____DbConn = pDBConn

            REM try load first time
            Me.reloadClass(pSQL)

            REM If no error then me.LoadID
            Me.LoadID(pTargettedRowID)

        End Sub

#End Region


#Region "Partial Access"

        ' Partial Simply means initial data is loaded directly from user but DBConn might be provided for reloadClass function to work

        ''' <summary>
        ''' Partial Access
        ''' </summary>
        ''' <param name="pDBConn"></param>
        ''' <param name="pTableRows"></param>
        ''' <param name="pTargettedRowID"></param>
        ''' <remarks></remarks>
        Sub New(
                ByVal pTableRows As DataRowCollection,
                      ByVal pTargettedRowID As Int64,
                     pTableName As String,
                     ByVal pDBConn As All__DBs)

            Me.New(pTableName)
            Me._OpenAccessFor = RecordAccessibility.PARTIAL_ACCESS
            Me._____DbConn = DBConn


            If pTableRows IsNot Nothing And pTableRows.Count > 0 Then

                Try

                    Dim dRows As IEnumerable(Of DataRow) = From d As DataRow In pTableRows.Cast(Of DataRow)()
                                                           Select d

                    Me.LoadFromRows(dRows)
                    If pTargettedRowID <> DO__NOT____TARGET__ANY_ROWID Then Me.LoadID(pTargettedRowID)

                Catch ex As Exception
                    Logger.Print("Error Converting from DatarowCollection to Table", ex)
                End Try

            End If
        End Sub


        ''' <summary>
        ''' Partial Access
        ''' </summary>
        ''' <param name="pDBConn"></param>
        ''' <param name="pTableRows"></param>
        ''' <param name="pTargettedRowID"></param>
        ''' <remarks></remarks>
        Sub New(
                ByVal pTableRows As IEnumerable(Of DataRow),
                      ByVal pTargettedRowID As Int64,
                      ByVal pTableName As String,
                     ByVal pDBConn As All__DBs)

            Me.New(pTableName)
            Me._OpenAccessFor = RecordAccessibility.PARTIAL_ACCESS
            Me._____DbConn = DBConn


            If pTableRows IsNot Nothing AndAlso pTableRows.Count > 0 Then

                Try
                    Me.LoadFromRows(pTableRows)
                    If pTargettedRowID <> DO__NOT____TARGET__ANY_ROWID Then Me.LoadID(pTargettedRowID)

                Catch ex As Exception
                    Logger.Print("Error Converting from IEnumerable(of datarow) to Table", ex)
                End Try

            End If
        End Sub

        ''' <summary>
        ''' Partial Access
        ''' </summary>
        ''' <param name="DBConn"></param>
        ''' <param name="FullTable"></param>
        ''' <param name="pTargettedRowID"></param>
        ''' <remarks></remarks>
        Sub New(
                ByVal FullTable As DataTable,
                      ByVal pTargettedRowID As Int64,
                      ByVal TableName As String,
                      ByVal DBConn As All__DBs)
            Me.New(TableName)
            Me._OpenAccessFor = RecordAccessibility.PARTIAL_ACCESS
            Me._____DbConn = DBConn


            If FullTable IsNot Nothing AndAlso FullTable.Rows.Count > 0 Then

                Try
                    Me.LoadFromRows(FullTable.Rows.Cast(Of DataRow))
                    If pTargettedRowID <> DO__NOT____TARGET__ANY_ROWID Then Me.LoadID(pTargettedRowID)

                Catch ex As Exception
                    Logger.Print("Error Creating Class From Table", ex)
                End Try

            Else
                '   Incase there is empty table
                Me.___RawTable = FullTable
            End If


            If Me.___RawTable IsNot Nothing Then Me.___RawTable.TableName = Me.____TableName

        End Sub


#End Region


#Region "Shallow Access"
        '   In the real definition, shallow reference partial. Just that it means partial with no DBConn

        ''' <summary>
        ''' Shallow Access. Should always target this passed in
        ''' </summary>
        ''' <param name="TargettedRow"></param>
        ''' <remarks>Assuming this Row has a Column Called ID and It will be assigned</remarks>
        Public Sub New(ByVal TargettedRow As DataRow)
            Me.New(New DataRow() {TargettedRow})
            If Me.hasRows() Then Me.TargettedRow_Cached = Me.AllRows(0)
        End Sub

        ''' <summary>
        ''' Shallow Access
        ''' </summary>
        ''' <param name="pTableRows"></param>
        ''' <remarks>Assuming this Row has a Column Called ID and It will be assigned</remarks>
        Public Sub New(ByVal pTableRows As IEnumerable(Of DataRow))
            Me.New(pTableRows, DO__NOT____TARGET__ANY_ROWID)
        End Sub

        ''' <summary>
        ''' Shallow Access
        ''' </summary>
        ''' <param name="pTableRows"></param>
        ''' <remarks>Assuming this Row has a Column Called ID and It will be assigned</remarks>
        Public Sub New(ByVal pTableRows As IEnumerable(Of DataRow),
                ByVal pTargettedRowID As Int64
                )
            Me.New(pTableRows, pTargettedRowID, Nothing)
        End Sub

        ''' <summary>
        ''' Shallow Access
        ''' </summary>
        ''' <param name="pTableRows"></param>
        ''' <remarks>Assuming this Row has a Column Called ID and It will be assigned</remarks>
        Public Sub New(ByVal pTableRows As IEnumerable(Of DataRow),
                ByVal pTargettedRowID As Int64,
                ByVal pTableName As String
                )
            Me.New(pTableRows, pTargettedRowID, pTableName, Nothing)
            Me._OpenAccessFor = RecordAccessibility.SHALLOW_ACCESS
        End Sub

        ''' <summary>
        ''' Shallow Access
        ''' </summary>
        ''' <remarks></remarks>
        Sub New(
                ByVal FullTable As DataTable,
                ByVal pTableName As String)

            Me.New(FullTable, pTableName, DO__NOT____TARGET__ANY_ROWID)
        End Sub

        ''' <summary>
        ''' Shallow Access
        ''' </summary>
        ''' <remarks></remarks>
        Sub New(
                ByVal FullTable As DataTable,
                ByVal pTableName As String,
                  ByVal pTargettedRowID As Int64)

            Me.New(FullTable, pTargettedRowID, pTableName, Nothing)
        End Sub

        ''' <summary>
        ''' Shallow Access
        ''' </summary>
        ''' <remarks></remarks>
        Sub New(
                ByVal pDataRows As DataRowCollection,
                ByVal pTableName As String)

            Me.New(pDataRows, DO__NOT____TARGET__ANY_ROWID, pTableName, Nothing)
            Me._OpenAccessFor = RecordAccessibility.SHALLOW_ACCESS
        End Sub

        ''' <summary>
        ''' Shallow Access
        ''' </summary>
        ''' <param name="pTableRows"></param>
        ''' <remarks>Assuming this Row has a Column Called ID and It will be assigned</remarks>
        Public Sub New(ByVal pTableRows As IEnumerable(Of DataRow),
                ByVal pTableName As String
                )
            Me.New(pTableRows, DO__NOT____TARGET__ANY_ROWID, pTableName)
        End Sub

        ''' <summary>
        ''' Shallow Access. Should always target this passed in
        ''' </summary>
        ''' <param name="pTargettedRow"></param>
        ''' <remarks>Assuming this Row has a Column Called ID and It will be assigned</remarks>
        Sub New(ByVal pTargettedRow As DataRow, ByVal pTableName As String)
            Me.New(pTargettedRow)
            Me.____TableName = pTableName
        End Sub

#End Region



#End Region

#Region "Properties"

        Public MustOverride ReadOnly Property ID As Int64

        Public MustOverride ReadOnly Property NumericPrimaryKeyName As String



        Protected LastSQL_Passed As String = String.Empty

        ''' <summary>
        ''' Database Connection
        ''' </summary>
        ''' <remarks></remarks>
        Private _____DbConn As All__DBs
        Public ReadOnly Property DBConn As All__DBs
            Get
                Return _____DbConn
            End Get
        End Property

        ''' <summary>
        ''' Only checks if numeric primary key column name is returned by this class not its physical presence in the current data rows
        ''' </summary>
        ''' <returns></returns>
        Public ReadOnly Property hasNumericPrimaryKeyColumn As Boolean
            Get
                Return Objects.EStrings.valueOf(Me.NumericPrimaryKeyName) <> String.Empty
            End Get
        End Property


        ''' <summary>
        ''' it is reloadable if db con is valid
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property isDBConnValid As Boolean
            Get
                Return Me._____DbConn IsNot Nothing
            End Get
        End Property

        Private ____TableName As String
        Public ReadOnly Property TableName As String
            Get
                Return Me.____TableName
            End Get
        End Property



        ''' <summary>
        ''' It returns datarowcollection format of this table
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property ClassData As DataRowCollection
            Get
                If Me.AllRows Is Nothing Then Return Nothing
                Return Me.RawTable.Rows
            End Get
        End Property


        Private Property ___RawTable As DataTable
        ''' <summary>
        ''' It should always contain data. if there is data at all
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property RawTable As DataTable
            Get
                Return Me.___RawTable
            End Get
        End Property



        ''' <summary>
        ''' For Full Access and Partial. Only for caching selected row
        ''' </summary>
        ''' <remarks></remarks>
        Private TargettedRow_Cached As DataRow = Nothing

        ''' <summary>
        ''' Use this always to receive property
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Overridable ReadOnly Property TargettedRow As DataRow
            Get

                Return Me.TargettedRow_Cached


            End Get
        End Property


        Private _OpenAccessFor As RecordAccessibility
        Public ReadOnly Property OpenAccessFor As RecordAccessibility
            Get
                Return Me._OpenAccessFor
            End Get
            'Set(ByVal value As RecordAccessibility)
            '    Me._OpenAccessFor = value
            'End Set
        End Property


        ''' <summary>
        ''' It always returns the enumerated table
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property AllRows As IEnumerable(Of DataRow)
            Get
                If Me.RawTable Is Nothing OrElse Me.RawTable.Rows.Count = 0 Then Return Nothing
                Return From d As DataRow In Me.RawTable.AsEnumerable()
                       Select d

            End Get
        End Property


        Public ReadOnly Property RowCount As Int32
            Get
                If Me.hasRows() Then Return Me.AllRows.Count()
                Return 0
            End Get
        End Property

#End Region


#Region "Methods"


        Friend Function LoadGeneric(Of T As {DB.Abstracts.SimpleTablePlugIn, New})(pTableName As String, pDataRow As DataRow, sender As T) As T
            Me.____TableName = pTableName
            Me.LoadFromRows(New DataRow() {pDataRow})
            If Me.hasRows() Then Me.TargettedRow_Cached = Me.AllRows(0)
            Me._OpenAccessFor = RecordAccessibility.SHALLOW_ACCESS
            Return sender
        End Function

        Private Sub LoadFromRows(pTableRows As IEnumerable(Of DataRow))
            Try
                Me.___RawTable = pTableRows.CopyToDataTable()
            Catch ex As Exception
                Logger.Print("Error Converting Rows to Table", ex)
            End Try
        End Sub


        ''' <summary>
        ''' Select from the pool of rows loaded the column with ID=pval
        ''' </summary>
        ''' <param name="pTargettedRowID"></param>
        ''' <remarks></remarks>
        Public Sub LoadID(pTargettedRowID As Int64)
            Try

                Me.TargettedRow_Cached = Nothing

                If Me.isTableValid AndAlso pTargettedRowID <> DO__NOT____TARGET__ANY_ROWID _
                    AndAlso Me.hasNumericPrimaryKeyColumn AndAlso Me.RawTable.Columns.Contains(Me.NumericPrimaryKeyName) Then


                    Dim pMatch As IEnumerable(Of DataRow) = From d As DataRow In Me.AllRows
                                                            Where ELong.valueOf(d(Me.NumericPrimaryKeyName)) = pTargettedRowID
                                                            Select d


                    If pMatch.Count() > 0 Then Me.TargettedRow_Cached = pMatch(0)

                End If

            Catch ex As Exception
                Logger.Print("Selecting ID: " & pTargettedRowID, ex)
            End Try
        End Sub


        ''' <summary>
        ''' Checks if the Table is valid. In other contains any rows
        ''' </summary>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Protected Overridable Function isTableValid() As Boolean
            Return Me.ClassData IsNot Nothing
        End Function

        Protected Function isTargettedRowValid() As Boolean
            Return Me.TargettedRow IsNot Nothing
        End Function


        ''' <summary>
        ''' Use this only if you have valid Connection
        ''' </summary>
        ''' <remarks></remarks>
        Public Sub reloadClass_usingLast_Passed_SQL()
            Me.reloadClass(Me.LastSQL_Passed)
        End Sub

        ''' <summary>
        ''' Reloads the class with your supplied SQL or else uses SELECT * FROM TABLE. Throws exception
        ''' </summary>
        ''' <param name="SQL">This sql is used to load the class if provided else it will select all from the table</param>
        ''' <remarks></remarks>
        Public Overridable Sub reloadClass(Optional ByVal SQL As String = vbNullString)
            Me.___RawTable = Nothing
            Me.TargettedRow_Cached = Nothing

            Try

                If Me.DBConn Is Nothing Then Return

                Dim ds As DataSet
                If SQL = vbNullString OrElse SQL.Equals(String.Empty) Then
                    ds = Me.DBConn.GetRS("SELECT * FROM " & Me.TableName)
                Else
                    ds = Me.DBConn.GetRS(SQL)
                End If

                If All__DBs.IsDataSetValid(ds) Then
                    Me.___RawTable = ds.Tables(0)

                End If


            Catch ex As Exception
                Logger.Print(ex)
            End Try


        End Sub

        Public Overridable Function hasRows() As Boolean
            Return Me.ClassData IsNot Nothing
        End Function

#End Region






    End Class

End Namespace

File: EEntityCore.DB/EEntityCore.DB/Abstracts/SimpleTablePlugIn.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using static EEntityCore.DB.Abstracts.All__DBs;
using ELibrary.Standard.VB.Objects;
using Microsoft.VisualBasic;

namespace EEntityCore.DB.Abstracts
{


    /// <summary>
    /// Simple Table with target row capability. Entity Template for Generators
    /// </summary>
    /// <remarks></remarks>
    public abstract class SimpleTablePlugIn
    {

        #region Enums & Consts

        public enum RecordAccessibility
        {
            /// <summary>
            /// Access table with SQL QUERIES and target record. reloadable with DBConn Valid
            /// </summary>
            /// <remarks></remarks>
            FULL_ACCESS,
            /// <summary>
            /// Access table and target record via direct record (DataRow(s)). reloadable if dbConn is valid
            /// </summary>
            /// <remarks></remarks>
            PARTIAL_ACCESS,
            // '' ''' <summary>
            // '' ''' Access table and target record via direcet record. reloadable
            // '' ''' </summary>
            // '' ''' <remarks></remarks>
            // ''PARTIAL_RELOADABLE_ACCESS
            // '' ''' <summary>
            // '' ''' Access table and target record. NOT reloadable [Cached]
            // '' ''' </summary>
            // '' ''' <remarks></remarks>
            // ''PARTIAL_FIXED_ACCESS
            /// <summary>
            /// Access target record only. NOT reloadable. No DBConn
            /// </summary>
            /// <remarks></remarks>
            SHALLOW_ACCESS,

            /// <summary>
            /// Signifies EMPTY UNDECIDED CLASS
            /// </summary>
            /// <remarks></remarks>
            UNKNOWN
        }

        public const int DO__NOT____TARGET__ANY_ROWID = 0;

        #endregion

        #region Contructors

        public SimpleTablePlugIn()
        {
            _OpenAccessFor = RecordAccessibility.UNKNOWN;
        }


        #region Full Access

        // Full Access means initial data is loaded directly from database, so DBConn MUST be provided

        /// <summary>
        /// FULL ACCESS. This will load data with dbconn
        /// </summary>
        /// <param name="DBConn"></param>
        /// <param name="TableName"></param>
        /// <remarks></remarks>
        public SimpleTablePlugIn(All__DBs DBConn) : this(DBConn, DO__NOT____TARGET__ANY_ROWID)
        {
        }

        /// <summary>
        /// Full Access. This will load data with dbconn
        /// </summary>
        /// <param name="DBConn"></param>
        /// <param name="pTableName"></param>
        /// <param name="pTargettedRowID">Only works if the table contains a column named ID</param>
        /// <remarks></remarks>
        public SimpleTablePlugIn(All__DBs DBConn, long pTargettedRowID)
        {
            _OpenAccessFor = RecordAccessibility.FULL_ACCESS;
            _____DbConn = DBConn;

            // REM try load first time
            ReloadClass("SELECT * FROM " + TableName);

            // REM If no error then me.LoadID
            LoadID(pTargettedRowID);
        }

        /// <summary>
        /// Full Access. This will load data with dbconn
        /// </summary>
        /// <param name="pDBConn"></param>
        /// <param name="pTableName"></param>
        /// <param name="pSQL">Load table with your own SQL</param>
        /// <remarks></remarks>
        public SimpleTablePlugIn(All__DBs pDBConn, string pSQL) : this(pDBConn, pSQL, DO__NOT____TARGET__ANY_ROWID)
        {
        }

        /// <summary>
        /// Full Access. This will load data with dbconn
        /// </summary>
        /// <param name="pDBConn"></param>
        /// <param name="pTableName"></param>
        /// <param name="pTargettedRowID"></param>
        /// <param name="pSQL">Load table with your own SQL</param>
        /// <remarks></remarks>
        public SimpleTablePlugIn(All__DBs pDBConn, string pSQL, long pTargettedRowID)
        {
            _OpenAccessFor = RecordAccessibility.FULL_ACCESS;
            _____DbConn = pDBConn;

            // REM try load first time
            ReloadClass(pSQL);

            // REM If no error then me.LoadID
            LoadID(pTargettedRowID);
        }

        #endregion


        #region Partial Access

        // Partial Simply means initial data is loaded directly from user but DBConn might be provided for reloadClass function to work
        // Also can have more than one row
        /// <summary>
        /// Partial Access
        /// </summary>
        /// <param name="pTableRows"></param>
        /// <param name="pTargettedRowID"></param>
        /// <remarks></remarks>
        public SimpleTablePlugIn(DataRowCollection pTableRows, long pTargettedRowID) : this(pTableRows: pTableRows.Cast<DataRow>(), pTargettedRowID: pTargettedRowID)
        {
        }


        /// <summary>
        /// Partial Access
        /// </summary>
        /// <param name="pTableRows"></param>
        /// <param name="pTargettedRowID"></param>
        /// <remarks></remarks>
        public SimpleTablePlugIn(IEnumerable<DataRow> pTableRows) : this(pTableRows, DO__NOT____TARGET__ANY_ROWID)
        {
        }


        /// <summary>
        /// Partial Access
        /// </summary>
        /// <param name="pTableRows"></param>
        /// <param name="pTargettedRowID"></param>
        /// <remarks></remarks>
        public SimpleTablePlugIn(IEnumerable<DataRow> pTableRows, long pTargettedRowID)
        {
            _OpenAccessFor = RecordAccessibility.PARTIAL_ACCESS;
            if (pTableRows is object && pTableRows.Count() > 0)
            {
                try
                {
                    LoadFromRows(pTableRows);
                    if (pTargettedRowID != DO__NOT____TARGET__ANY_ROWID)
                        LoadID(pTargettedRowID);
                }
                catch (Exception ex)
                {
                    Logger.Print("Error Converting from IEnumerable(of datarow) to Table", ex);
                }
            }
        }

        /// <summary>
        /// Partial Access
        /// </summary>
        /// <param name="FullTable"></param>
        /// <param name="pTargettedRowID"></param>
        /// <remarks></remarks>
        public SimpleTablePlugIn(DataTable FullTable, long pTargettedRowID)
        {
            _OpenAccessFor = RecordAccessibility.PARTIAL_ACCESS;

            // Incase there is empty table
            ___RawTable = FullTable;
            if (FullTable is object && FullTable.Rows.Count > 0)
            {
                LoadID(pTargettedRowID);
            }
        }


        #endregion


        #region Shallow Access
        // In the real definition, shallow reference partial. Just that it means partial with no DBConn
        // Shallow is always one row
        /// <summary>
        /// Shallow Access. Should always target this passed in
        /// </summary>
        /// <param name="TargettedRow"></param>
        /// <remarks>Assuming this Row has a Column Called ID and It will be assigned</remarks>
        public SimpleTablePlugIn(DataRow TargettedRow) : this(new DataRow[] { TargettedRow }, pTargettedRowID: DO__NOT____TARGET__ANY_ROWID)
        {
            _OpenAccessFor = RecordAccessibility.SHALLOW_ACCESS;
            if (HasRows())
                TargettedRow_Cached = AllRows.ElementAtOrDefault(0);
        }

        #endregion



        #endregion

        #region Properties

        /// <summary>
        /// Note: your table must have the column name ID for you to call this
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public virtual long ID
        {
            get
            {
                if (!IsTargettedRowValid)
                    return 0L;
                return ELong.ValueOf(TargettedRow[NumericPrimaryKeyName]);
            }
        }

        public virtual string NumericPrimaryKeyName
        {
            get
            {
                return "ID";
            }
        }

        protected string LastSQL_Passed = string.Empty;

        /// <summary>
        /// Database Connection
        /// </summary>
        /// <remarks></remarks>
        private All__DBs _____DbConn;

        public All__DBs DBConn
        {
            get
            {
                return _____DbConn;
            }
        }

        /// <summary>
        /// Only checks if numeric primary key column name is returned by this class not its physical presence in the current data rows
        /// </summary>
        /// <returns></returns>
        public bool HasNumericPrimaryKeyColumn
        {
            get
            {
                return (EStrings.ValueOf(NumericPrimaryKeyName) ?? "") != (string.Empty ?? "");
            }
        }


        /// <summary>
        /// it is reloadable if db con is valid
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public bool IsDBConnValid
        {
            get
            {
                return _____DbConn is object;
            }
        }

        public abstract string TableName { get; }
        private DataTable ___RawTable { get; set; }
        /// <summary>
        /// It should always contain data. if there is data at all
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public DataTable RawTable
        {
            get
            {
                return ___RawTable;
            }
        }



        /// <summary>
        /// For Full Access and Partial. Only for caching selected row
        /// </summary>
        /// <remarks></remarks>
        private DataRow TargettedRow_Cached = null;

        /// <summary>
        /// Use this always to receive property. TargettedRow_Cached
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public virtual DataRow TargettedRow
        {
            get
            {
                return TargettedRow_Cached;
            }
        }

        private RecordAccessibility _OpenAccessFor;

        public RecordAccessibility OpenAccessFor
        {
            get
            {
                return _OpenAccessFor;
            }
            // Set(ByVal value As RecordAccessibility)
            // Me._OpenAccessFor = value
            // End Set
        }


        /// <summary>
        /// It always returns the enumerated table
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public IEnumerable<DataRow> AllRows
        {
            get
            {
                if (RawTable is null || RawTable.Rows.Count == 0)
                    return null;
                return from d in RawTable.AsEnumerable()
                       select d;
            }
        }

        public int RowCount
        {
            get
            {
                if (HasRows())
                    return AllRows.Count();
                return 0;
            }
        }

        #endregion


        #region Methods


        internal T LoadGeneric<T>(DataRow pDataRow) where T : SimpleTablePlugIn, new()
        {
            LoadFromRows(new DataRow[] { pDataRow });
            if (HasRows())
                TargettedRow_Cached = AllRows.ElementAtOrDefault(0);
            _OpenAccessFor = RecordAccessibility.SHALLOW_ACCESS;
            return (T)this;
        }

        private void LoadFromRows(IEnumerable<DataRow> pTableRows)
        {
            try
            {
                ___RawTable = pTableRows.CopyToDataTable();
            }
            catch (Exception ex)
            {
                Logger.Print("Error Converting Rows to Table", ex);
            }
        }


        /// <summary>
        /// Select from the pool of rows loaded the column with ID=pval
        /// </summary>
        /// <param name="pTargettedRowID"></param>
        /// <remarks></remarks>
        public void LoadID(long pTargettedRowID)
        {
            try
            {
                TargettedRow_Cached = null;
                if (IsTableValid() && pTargettedRowID != DO__NOT____TARGET__ANY_ROWID && HasNumericPrimaryKeyColumn && RawTable.Columns.Contains(NumericPrimaryKeyName))
                {
                    var pMatch = from d in AllRows
                                 where ELong.ValueOf(d[NumericPrimaryKeyName]) == pTargettedRowID
                                 select d;
                    if (pMatch.Count() > 0)
                        TargettedRow_Cached = pMatch.ElementAtOrDefault(0);
                }
            }
            catch (Exception ex)
            {
                Logger.Print("Selecting ID: " + pTargettedRowID, ex);
            }
        }


        /// <summary>
        /// Checks if the Table is valid. In other contains any rows
        /// </summary>
        /// <returns></returns>
        /// <remarks></remarks>
        protected virtual bool IsTableValid()
        {
            return HasRows();
        }

        public bool IsTargettedRowValid
        {
            get
            {
                return TargettedRow is object;
            }
        }


        /// <summary>
        /// Use this only if you have valid Connection
        /// </summary>
        /// <remarks></remarks>
        public void ReloadClass_usingLast_Passed_SQL()
        {
            ReloadClass(LastSQL_Passed);
        }


        /// <summary>
        /// Reloads the class with your supplied SQL or else uses SELECT * FROM TABLE. Throws exception. Reloads class and retarget class
        /// </summary>
        /// <param name="SQL">This sql is used to load the class if provided else it will select all from the table</param>
        /// <remarks></remarks>
        public virtual void ReloadClass(string SQL = Constants.vbNullString)
        {
            ___RawTable = null;
            TargettedRow_Cached = null;
            long pLastID = ID;
            if (DBConn is null)
                return;

            DataSet ds;
            if (string.IsNullOrEmpty(SQL) || SQL.Equals(string.Empty))
            {
                ds = DBConn.GetRS("SELECT * FROM " + TableName);
            }
            else
            {
                ds = DBConn.GetRS(SQL);
            }

            if (IsDataSetValid(ds))
            {
                ___RawTable = ds.Tables[0];
            }

            // Retarget if possible
            if (pLastID != DO__NOT____TARGET__ANY_ROWID)
                LoadID(pLastID);
        }

        public virtual bool HasRows()
        {
            if (RawTable is null)
                return false;
            return RawTable.Rows.Count > 0;
        }



        /// <summary>
        /// Deletes a row using an id. *Already functional. if DBConn is valid
        /// </summary>
        /// <param name="pID"></param>
        /// <returns></returns>
        /// <remarks></remarks>
        public virtual bool DeleteRow(long pID)
        {
            if (DBConn is null)
                throw new Exception("DbConn was not set for this class. Class Load Status: " + OpenAccessFor.ToString());
            return DeleteRow(pID: pID, pDBConn: DBConn, pTableName: TableName);
        }

        /// <summary>
        /// Deletes a row using an id. *Already functional. Only if dbConn is Available
        /// </summary>
        /// <returns></returns>
        /// <remarks></remarks>
        public virtual bool DeleteRow()
        {
            return DeleteRow(ID);
        }

        public static bool DeleteRow(All__DBs pDBConn, long pID, string pTableName)
        {
            try
            {
                if (pDBConn is null)
                    throw new Exception("DbConn Is NOT VALID: ");
                return pDBConn.DbExec(string.Format("DELETE FROM {0} WHERE ID={1} ", pTableName, pID)) >= 1;
            }
            catch (Exception )
            {
                throw;
            }

        }


        #endregion






    }
}